File: src/papyrus_api/routers/notes.py
"""笔记管理API路由。"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from papyrus.data.notes_storage import (
    Note,
    load_notes,
    create_note,
    update_note,
    delete_note as delete_note_storage,
)
from papyrus.integrations.obsidian import sync_obsidian_vault
from papyrus.paths import NOTES_FILE

router = APIRouter(prefix="/notes", tags=["notes"])


class NoteDict(BaseModel):
    id: str
    title: str
    folder: str
    content: str
    preview: str
    tags: list[str]
    created_at: float
    updated_at: float
    word_count: int
    hash: str = ""
    headings: list[dict[str, Any]] = []
    outgoing_links: list[str] = []
    incoming_count: int = 0


class NotesListResponse(BaseModel):
    success: bool
    notes: list[NoteDict]
    count: int


class CreateNoteIn(BaseModel):
    title: str
    folder: str = "默认"
    content: str = ""
    tags: list[str] = []


class CreateNoteResponse(BaseModel):
    success: bool
    note: NoteDict


class UpdateNoteIn(BaseModel):
    title: str | None = None
    folder: str | None = None
    content: str | None = None
    tags: list[str] | None = None


class UpdateNoteResponse(BaseModel):
    success: bool
    note: NoteDict


class DeleteNoteResponse(BaseModel):
    success: bool


class ObsidianImportIn(BaseModel):
    vault_path: str
    exclude_folders: list[str] = [".obsidian", ".git"]


class ObsidianImportResponse(BaseModel):
    success: bool
    imported: int
    skipped: int
    errors: list[str]


def _note_to_dict(note: Note) -> NoteDict:
    """将Note对象转换为NoteDict。"""
    return NoteDict(
        id=note.id,
        title=note.title,
        folder=note.folder,
        content=note.content,
        preview=note.preview,
        tags=note.tags,
        created_at=note.created_at,
        updated_at=note.updated_at,
        word_count=note.word_count,
        hash=getattr(note, "hash", ""),
        headings=getattr(note, "headings", []),
        outgoing_links=getattr(note, "outgoing_links", []),
        incoming_count=getattr(note, "incoming_count", 0),
    )


@router.get("", response_model=NotesListResponse)
def list_notes() -> NotesListResponse:
    """获取所有笔记列表。"""
    notes = load_notes(NOTES_FILE)
    return NotesListResponse(
        success=True,
        notes=[_note_to_dict(n) for n in notes],
        count=len(notes),
    )


@router.post("", response_model=CreateNoteResponse)
def create_note_endpoint(payload: CreateNoteIn) -> CreateNoteResponse:
    """创建新笔记。"""
    note = create_note(
        NOTES_FILE,
        title=payload.title,
        folder=payload.folder,
        content=payload.content,
        tags=payload.tags,
    )
    return CreateNoteResponse(success=True, note=_note_to_dict(note))


@router.get("/{note_id}", response_model=CreateNoteResponse)
def get_note_endpoint(note_id: str) -> CreateNoteResponse:
    """获取单个笔记详情。"""
    notes = load_notes(NOTES_FILE)
    for note in notes:
        if note.id == note_id:
            return CreateNoteResponse(success=True, note=_note_to_dict(note))
    raise HTTPException(status_code=404, detail="note not found")


@router.patch("/{note_id}", response_model=UpdateNoteResponse)
def update_note_endpoint(note_id: str, payload: UpdateNoteIn) -> UpdateNoteResponse:
    """更新笔记。"""
    note = update_note(
        NOTES_FILE,
        note_id=note_id,
        title=payload.title,
        folder=payload.folder,
        content=payload.content,
        tags=payload.tags,
    )
    if note is None:
        raise HTTPException(status_code=404, detail="note not found")
    return UpdateNoteResponse(success=True, note=_note_to_dict(note))


@router.delete("/{note_id}", response_model=DeleteNoteResponse)
def delete_note_endpoint(note_id: str) -> DeleteNoteResponse:
    """删除笔记。"""
    ok = delete_note_storage(NOTES_FILE, note_id)
    if not ok:
        raise HTTPException(status_code=404, detail="note not found")
    return DeleteNoteResponse(success=True)


@router.post("/import/obsidian", response_model=ObsidianImportResponse)
def import_obsidian_endpoint(payload: ObsidianImportIn) -> ObsidianImportResponse:
    """从Obsidian Vault导入笔记。"""
    import os
    from pathlib import Path

    vault = Path(payload.vault_path).resolve()
    if not vault.exists():
        raise HTTPException(status_code=400, detail="Vault path does not exist")
    
    # SECURITY: basic path traversal check
    try:
        # Ensure the path looks like a reasonable vault directory
        if not vault.is_dir():
            raise HTTPException(status_code=400, detail="Vault path is not a directory")
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid vault path")

    existing_notes = load_notes(NOTES_FILE)

    result = sync_obsidian_vault(
        NOTES_FILE,
        payload.vault_path,
        existing_notes,
        exclude_folders=payload.exclude_folders,
    )

    return ObsidianImportResponse(
        success=True,
        imported=result.imported,
        skipped=result.skipped,
        errors=result.errors,
    )

File: backend/src/api/routes/notes.ts
import type { FastifyInstance } from 'fastify';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { getAllNotes, createNote, updateNote, deleteNote, deleteNotes, getNoteById } from '../../core/notes.js';
import { importObsidianVault } from '../../core/notes.js';
import { recordNoteCreated } from '../../core/progress.js';
import { pushExtensionEvent } from '#/core/extension-events.js';

export default async function notesRoutes(fastify: FastifyInstance): Promise<void> {
  fastify.get('/', async (_request, reply) => {
    const notes = getAllNotes();
    reply.send({ success: true, notes, count: notes.length });
  });

  fastify.post('/', async (request, reply) => {
    try {
      const body = request.body as { title: string; folder?: string; content?: string; tags?: string[] };
      if (!body.title) {
        reply.status(400).send({ success: false, error: 'Title is required' });
        return;
      }
      const note = createNote(body.title, body.content ?? '', body.folder, body.tags ?? []);
      recordNoteCreated();
      reply.send({ success: true, note });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.get('/:noteId', async (request, reply) => {
    const { noteId } = request.params as { noteId: string };
    const note = getNoteById(noteId);
    if (!note) {
      reply.status(404).send({ success: false, error: 'Note not found' });
      return;
    }
    reply.send({ success: true, note });
  });

  fastify.patch('/:noteId', async (request, reply) => {
    try {
      const { noteId } = request.params as { noteId: string };
      const body = request.body as { title?: string; folder?: string; content?: string; tags?: string[] };
      const note = await updateNote(noteId, body);
      if (!note) {
        reply.status(404).send({ success: false, error: 'Note not found' });
        return;
      }
      pushExtensionEvent('note.modified', {
        note_id: noteId,
        title: note.title,
        updated_at: note.updated_at,
      });
      reply.send({ success: true, note });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.delete('/:noteId', async (request, reply) => {
    const { noteId } = request.params as { noteId: string };
    const success = deleteNote(noteId);
    if (!success) {
      reply.status(404).send({ success: false, error: 'Note not found' });
      return;
    }
    reply.send({ success: true });
  });

  fastify.post('/batch-delete', async (request, reply) => {
    const body = request.body as { ids?: string[] };
    if (!body.ids || body.ids.length === 0) {
      reply.status(400).send({ success: false, error: 'ids is required' });
      return;
    }
    const deleted = deleteNotes(body.ids);
    reply.send({ success: true, deleted });
  });

  fastify.post('/import/obsidian', async (request, reply) => {
    const body = request.body as { vault_path?: string; exclude_folders?: string[] };
    if (!body.vault_path) {
      reply.status(400).send({ success: false, error: 'vault_path is required' });
      return;
    }
    if (body.vault_path.includes('\x00')) {
      reply.status(400).send({ success: false, error: '非法路径' });
      return;
    }
    let resolvedVaultPath: string;
    try {
      resolvedVaultPath = fs.realpathSync(path.resolve(body.vault_path));
    } catch {
      reply.status(400).send({ success: false, error: '路径不存在或无法访问' });
      return;
    }
    const homeDir = fs.realpathSync(path.resolve(os.homedir()));
    if (!resolvedVaultPath.startsWith(homeDir + path.sep) && resolvedVaultPath !== homeDir) {
      reply.status(400).send({ success: false, error: 'Obsidian vault 路径必须在用户主目录下' });
      return;
    }
    try {
      const result = importObsidianVault(resolvedVaultPath, body.exclude_folders);
      if (result.error) {
        reply.status(400).send({ success: false, error: result.error, imported: 0, errors: 0 });
        return;
      }
      reply.send({ success: true, imported: result.imported, skipped: 0, errors: result.errors });
    } catch (err) {
      const message = err instanceof Error ? err.message : '导入过程中发生未知错误';
      reply.status(500).send({ success: false, error: message });
    }
  });
}
